### PR TITLE
Lester feature 2 sale reportmanager

### PR DESCRIPTION
**Pull request by tofu-lab**
I finished most of my feature and uploaded sample data for history that my feature will be pulling data from but there are a few small things:

- Need to connect my screen to the rest of the program (probably after @tonyxieuoft 's pull request is merged?)
- Need a way to change the filepath of the GatewayReader so that SalesReporterController can get the data from the correct file. @Robinrswanson I think a setFilePath or something should be added to the gateway interface? 
- Need @wuhanyi70 's feature so that SalesReporter can check if the dates entered are valid

### CODE REVIEW

File: src/main/java/interface_adaptors/sales_report/SalesReporterController.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    }
+
+    /**
+     * The controller packages the data and sends it off to the Update Use Case
+     * @param startTime the start of the time range for the sales report
+     * @param endTime the end of the time range for the sales report
+     */
+    public void reportData(String startTime, String endTime) throws IOException {
+        List<String[]> readerData = gatewayReader.getData();
+        salesReportUseCase.generateReport(readerData, startTime, endTime);
```
Comment (tonyxieuoft) [inline src/main/java/interface_adaptors/sales_report/SalesReporterController.java]: Before passing the data to the use case, maybe package it into a class named "SalesReporterInputData"? The class would just contain three instance variables: readerData, startTime, endTime (the parameters), and help adhere to clean architecture/remove potential code smells.

File: src/main/java/screens/SalesReportScreen.java [hunk 2]
```diff
@@ -1,82 +1,183 @@
...
 import java.awt.*;
 import java.awt.event.ActionEvent;
 import java.awt.event.ActionListener;
+import java.io.IOException;
 import java.util.ArrayList;
 
-public class SalesReportScreen extends JPanel implements ActionListener {
-    public static final int VISIBLE_ROWS = 10;
+
+public class SalesReportScreen extends JPanel implements SalesReporterView {
```
Comment (tonyxieuoft) [inline src/main/java/screens/SalesReportScreen.java:15]: The screen and the use case looks good!

File: src/main/java/use_cases/sales_report/SalesReporter.java [hunk 3]
```diff
@@ -7,32 +7,46 @@
...
+    public void generateReport(List<String[]> rows, String startTime, String endTime){
+        LocalDateTime start = stringToDateTime(startTime);
+        LocalDateTime end = stringToDateTime(endTime);
+        if (!checker.StartDateValid(start)){
+            presenter.prepareFailure(SalesReporterOutputBoundary.START_TIME_ERROR);
         }
-        String totalRevenue = String.format("Total Revenue: $%.5f", totalRevenue(splitData));
-        result.add(new String[]{totalRevenue});
+        else if (!checker.EndDateValid(end)){
+            presenter.prepareFailure(SalesReporterOutputBoundary.END_TIME_ERROR);
```
Comment (tonyxieuoft) [inline src/main/java/use_cases/sales_report/SalesReporter.java:29]: I can't seem to find the SalesReporterOutputBoundary interface or the presenter... not sure what the reason is though.

Comment (tonyxieuoft) [reply src/main/java/use_cases/sales_report/SalesReporter.java:29]: It might be that I'm just not on the most recent version.

Comment (tofu-lab) [reply src/main/java/use_cases/sales_report/SalesReporter.java:29]: Yeah i dont know why it wasn't pushed. SalesReporterView, SalesReporterOutputBoundary, and SalesReporterPresenter are all highlighted red in my project.
image

not sure why

File: src/main/java/interface_adaptors/sales_report/SalesReporterController.java (hunk 1 above)
Comment (tofu-lab) [reply src/main/java/interface_adaptors/sales_report/SalesReporterController.java]: ok, i'll look into this

**Review by Robinrswanson — COMMENTED** [review]
In terms of functionality everything looks good. For the SalesReporter, the class is pretty big and it might be better to try and extract some of the methods into other classes. You might be able to do this by creating a data structure for the ArrayList<String[]> and having it implement some of the methods.

File: src/main/java/interface_adaptors/sales_report/SalesReporterController.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+import use_cases.sales_report.SalesReporterInputData;
+
+import java.io.IOException;
+import java.util.List;
+
+public class SalesReporterController {
+    private final SalesReporterInputBoundary salesReportUseCase;
+    private final GatewayReaderInterface reader;
+    private final String FILE_PATH =
+            "C:\\Users\\leste\\OneDrive - University of Toronto\\School\\Fall 2022\\CSC 207\\course-project-group-131\\src\\main\\java\\database\\Sample Data - Sample History.csv";
```
Comment (Bobasuar123) [inline src/main/java/interface_adaptors/sales_report/SalesReporterController.java]: Maybe change this line to Path from Repository Root, (src\\main\\java\\database\\Sample Data - Sample History.csv)

**Pull request merged**